akka-cluster-sharding/src/main/scala/akka/cluster/sharding/dynamic/DynamicShardAllocationStrategy.scala

```diff
@@ -0,0 +1,221 @@
+        log.debug("Received updated shard locations [{}] all locations are now [{}]", newLocations, currentLocations)
+      case GetShardLocation(shard) =>
+        log.debug("GetShardLocation [{}]", shard)
+        val shardLocation = currentLocations.get(shard).map(asStr => AddressFromURIString(asStr))
+        sender() ! GetShardLocationResponse(shardLocation)
+      case GetShardLocations =>
+        log.debug("GetShardLocations")
+        sender() ! GetShardLocationsResponse(currentLocations.mapValues(asStr => AddressFromURIString(asStr)).toMap)
+    }
+  }
+}
+
+class DynamicShardAllocationStrategy(system: ActorSystem, typeName: String)(
+    // local only ask
+    implicit val timeout: Timeout = Timeout(5.seconds))
+    extends ShardCoordinator.StartableAllocationStrategy {
+
+  import DynamicShardAllocationStrategy._
+  import akka.pattern.ask
+  import system.dispatcher
+
+  private val log = Logging(system, classOf[DynamicShardAllocationStrategy])
+
+  private var shardState: ActorRef = _
+
+  private[akka] def createShardStateActor(): ActorRef = {
+    system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(DDataStateActor.props(typeName), s"dynamic-allocation-state-$typeName")
+  }
+
+  private val cluster = Cluster(system)
+
+  override def start(): Unit = {
+    shardState = createShardStateActor()
+  }
+
+  override def allocateShard(
+      requester: ShardRegion,
+      shardId: ShardId,
+      currentShardAllocations: Map[ShardRegion, immutable.IndexedSeq[ShardId]]): Future[ShardRegion] = {
+
+    log.debug("allocateShard [{}] [{}] [{}]", shardId, requester, currentShardAllocations)
+
+    // current shard allocations include all current shard regions
+    (shardState ? GetShardLocation(shardId))
+      .mapTo[GetShardLocationResponse]
+      .map {
+        case GetShardLocationResponse(None) =>
+          log.debug("No specific location for shard [{}]. Allocating to requester [{}]", shardId, requester)
+          requester
+        case GetShardLocationResponse(Some(address)) =>
+          // if it is the local address, convert it so it is found in the shards
+          if (address == cluster.selfAddress) {
+            currentShardAllocations.keys.find(_.path.address.hasLocalScope) match {
+              case None =>
+                log.warning("unable to find local shard in currentShardAllocation. Using requester")
+                requester
+              case Some(localShardRegion) =>
+                log.debug("allocating to local shard")
+                localShardRegion
+            }
+          } else {
+            currentShardAllocations.keys.find(_.path.address == address) match {
+              case None =>
+                log.warning(
+                  "Dynamic shard location [{}] for shard [{}[ not found in members [{}]",
+                  address,
+                  shardId,
+                  currentShardAllocations.keys.mkString(","))
+                requester
```

Comment:
Same race condition, but with other nodes joining an already-existing coordinator.

Once we've made these allocations, can a 'dynamic' shard allocation strategy trigger a rebalance?

Reply:
That isn't supported yet, we rebalance every 10s by default, I think that could be good enough for now?